src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/AbstractMultiLifetimeComponent.cs

```diff
@@ -11,33 +12,44 @@ namespace Microsoft.VisualStudio.ProjectSystem
     ///     An <see langword="abstract"/> base class that simplifies the lifetime of 
     ///     a component that is loaded and unloaded multiple times.
     /// </summary>
-    internal abstract class AbstractMultiLifetimeComponent : OnceInitializedOnceDisposedAsync
+    internal abstract class AbstractMultiLifetimeComponent<T> : OnceInitializedOnceDisposedAsync
+        where T : class, IMultiLifetimeInstance
     {
         private readonly object _lock = new object();
-        private TaskCompletionSource<object> _loadedSource = new TaskCompletionSource<object>();
-        private IMultiLifetimeInstance _instance;
+        private TaskCompletionSource<(T instance, JoinableTask initializeAsyncTask)> _instanceTaskSource = new TaskCompletionSource<(T instance, JoinableTask initializeAsyncTask)>();
 
         protected AbstractMultiLifetimeComponent(JoinableTaskContextNode joinableTaskContextNode)
-            : base(joinableTaskContextNode)
+             : base(joinableTaskContextNode)
         {
         }
 
-        public IMultiLifetimeInstance Instance
-        {
-            get { return _instance; }
-        }
-
         /// <summary>
-        ///     Gets a task that is completed when current <see cref="AbstractMultiLifetimeComponent"/> has 
-        ///     completed loading.
+        ///     Returns a task that will complete when current <see cref="AbstractMultiLifetimeComponent{T}"/> has completed
+        ///     loading and has published its instance.
         /// </summary>
+        /// <exception cref="OperationCanceledException">
+        ///     The result is awaited and the <see cref="ConfiguredProject"/> is unloaded.
+        ///     <para>
+        ///         -or
+        ///     </para>
+        ///     The result is awaited and <paramref name="cancellationToken"/> is cancelled.
+        /// </exception>
         /// <remarks>
-        ///     The returned <see cref="Task"/> is canceled when the <see cref="AbstractMultiLifetimeComponent"/> 
-        ///     is disposed.
+        ///     This method does not initiate loading of the <see cref="AbstractMultiLifetimeComponent{T}"/>, however,
+        ///     it will join the load when it starts.
         /// </remarks>
-        public Task Loaded
+        protected async Task<T> PublishInstanceAsync(CancellationToken cancellationToken = default)
```

Comment:
`PublishInstanceAsync` sounds like it initiates publish (a directive). Would `GetInstanceAsync` or `WhenPublished` be clearer?

Reply:
I changed this name, but left WorkspaceContextHost to be still be called `PublishAsync`. This follows the same pattern for other components that "publish", such as the tree, but really just waiting for the specified version to published.